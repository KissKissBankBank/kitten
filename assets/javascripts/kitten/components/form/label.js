import React from 'react'
import classNames from 'classnames'
import domElementHelper from 'kitten/helpers/dom/element-helper'

export class Label extends React.Component {
  constructor(props) {
    super(props)

    this.handleClick = this.handleClick.bind(this)
  }

  handleClick(e) {
    if (domElementHelper.canUseDom() && this.props.focusId) {
      e.preventDefault()

      document.getElementById(this.props.focusId).focus()
    }
  }

  render() {
    const {
      tag,
      className,
      children,
      focusId,
      size,
      withoutPointerEvents,
      ...other,
    } = this.props

    const Tag = tag

    const labelClassName = classNames(
      "k-Label",
      className,
      {
        [`k-Label--${size}`]: size,
        'k-Label--withoutPointerEvents': withoutPointerEvents,
      },
    )

    return (
      <Tag className={ labelClassName }
           htmlFor={ tag == 'label' ? focusId : null }
           onClick={ this.handleClick }
           { ...other }>
        { children }
      </Tag>
    )
  }
}

Label.defaultProps = {
  tag: 'label',
  className: null,
  children: 'Label',
  focusId: null,
  size: null, // `tiny`
  withoutPointerEvents: false,
}
