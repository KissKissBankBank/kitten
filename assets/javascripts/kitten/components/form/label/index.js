import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import classNames from 'classnames'
import isEmpty from 'lodash/fp/isEmpty'
import domElementHelper from '../../../helpers/dom/element-helper'
import TYPOGRAPHY from '../../../constants/typography-config'
import { pxToRem, stepToRem } from '../../../helpers/utils/typography'
import { checkDeprecatedSizes } from '../../../helpers/utils/deprecated'

const StyledLabel = styled.label`
  display: block;
  ${TYPOGRAPHY.fontStyles.regular}
  cursor: pointer;
  line-height: 1.3;

  &.k-Label--micro {
    font-size: ${stepToRem(-2)};
  }

  &.k-Label--small {
    font-size: ${stepToRem(-1)};
  }

  &.k-Label--medium {
    font-size: ${stepToRem(0)};
  }

  &.k-Label--withoutPointerEvents {
    pointer-events: none;
  }

  .k-Label--dot {
    margin: 0 0 0 ${pxToRem(10)};
    width: var(--dot-width);
    height: var(--dot-width);
    background-color: var(--dot-background-color);
    vertical-align: middle;
    display: inline-block;
    border-radius: var(--border-radius-rounded);
  }
`

export const Label = ({
  tag,
  className,
  children,
  focusId,
  size,
  withoutPointerEvents,
  htmlFor,
  dot,
  style,
  ...other
}) => {
  checkDeprecatedSizes(size)

  const handleClick = e => {
    if (domElementHelper.canUseDom() && focusId) {
      e.preventDefault()

      document.getElementById(focusId).focus()
    }
  }

  return (
    <StyledLabel
      className={classNames('k-Label', className, `k-Label--${size}`, {
        'k-Label--withoutPointerEvents': withoutPointerEvents,
      })}
      htmlFor={focusId || htmlFor}
      onClick={handleClick}
      as={tag}
      {...other}
    >
      {children}
      {!isEmpty(dot) && (
        <span
          className={classNames('k-Label--dot')}
          title={dot.title}
          tabIndex="-1"
          style={{
            '--dot-background-color': dot?.backgroundColor ?? null,
            '--dot-width': 'width' in dot ? pxToRem(dot.width) : null,
            ...style,
          }}
        />
      )}
    </StyledLabel>
  )
}

Label.defaultProps = {
  children: 'Label',
  focusId: null,
  size: 'small',
  withoutPointerEvents: false,
  dot: {},
}

Label.propTypes = {
  focusId: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  size: PropTypes.oneOf(['micro', 'small', 'medium']),
  withoutPointerEvents: PropTypes.bool,
  dot: PropTypes.shape({
    width: PropTypes.number,
    backgroundColor: PropTypes.node,
    title: PropTypes.string,
  }),
}
