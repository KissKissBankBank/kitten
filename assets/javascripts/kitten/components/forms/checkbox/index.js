import React, { useRef, useEffect } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import styled from 'styled-components'
import { Text } from '../../../components/typography/text'
import { pxToRem, stepToRem } from '../../../helpers/utils/typography'
import COLORS from '../../../constants/colors-config'
import TYPOGRAPHY from '../../../constants/typography-config'

const StyledCheckbox = styled.div`
  ${TYPOGRAPHY.fontStyles.light}
  font-size: ${stepToRem(-1)};
  line-height: 1.5;

  display: flex;
  align-items: flex-start;
  flex-wrap: wrap;

  margin: ${pxToRem(10)} 0;

  &.k-Form-Checkbox--disabled input,
  &.k-Form-Checkbox--disabled label {
    cursor: not-allowed;
  }

  .k-Form-Checkbox__label {
    flex-basis: calc(100% - ${pxToRem(20 + 10)});
    cursor: pointer;
    transition: color 0.2s;
    ${TYPOGRAPHY.fontStyles.light}

    &:active {
      color: ${COLORS.primary3};
    }
  }

  .k-Form-Checkbox__input {
    appearance: none;

    box-sizing: border-box;
    display: inline-block;
    flex-shrink: 0;
    width: ${pxToRem(20)};
    height: ${pxToRem(20)};
    margin: 0 ${pxToRem(10)} 0 0;
    border-radius: var(--border-radius-s);

    background-color: ${COLORS.background1};
    border: var(--border);
    cursor: pointer;
    transition: border-color 0.2s, background-color 0.2s;

    &:focus {
      border: var(--border-active-primary);
    }

    &:checked,
    &:indeterminate {
      background-color: ${COLORS.primary1};
      background-repeat: no-repeat;
      background-position: center;
      border: var(--border-active-primary);
    }

    &:checked {
      background-image: url("data:image/svg+xml,%3Csvg width='11' height='8' viewBox='0 0 11 8' xmlns='http://www.w3.org/2000/svg' fill='%23fff'%0A%3E%3Cpath d='M4.44 7.596L3.024 6.182 8.682.525l1.414 1.414z' /%3E%3Cpath d='M4.44 7.596L.903 4.06l1.414-1.414 3.536 3.536z' /%3E%3C/svg%3E%0A");
    }

    &:indeterminate {
      background-image: url("data:image/svg+xml,%3Csvg width='6' height='2' viewBox='0 0 6 2' xmlns='http://www.w3.org/2000/svg' fill='%23fff'%0A%3E%3Crect width='6' height='2' /%3E%3C/svg%3E%0A");
    }

    &:disabled {
      background-color: ${COLORS.line1};
      border: var(--border-disabled);
    }

    &:disabled:checked {
      border: var(--border-disabled-active);
      background-color: ${COLORS.line2};
      background-image: url("data:image/svg+xml,%3Csvg width='11' height='8' viewBox='0 0 11 8' xmlns='http://www.w3.org/2000/svg' fill='%23b8b8b8'%0A%3E%3Cpath d='M4.44 7.596L3.024 6.182 8.682.525l1.414 1.414z' /%3E%3Cpath d='M4.44 7.596L.903 4.06l1.414-1.414 3.536 3.536z' /%3E%3C/svg%3E%0A");
    }
  }

  &.k-Form-Checkbox--error .k-Form-Checkbox__input {
    border: var(--border-danger);

    &:checked {
      background-color: ${COLORS.error};

      &:focus {
        box-shadow: 0 0 ${pxToRem(3)} ${COLORS.error};
      }
    }
  }

  .k-Form-Checkbox__link,
  label a {
    ${TYPOGRAPHY.fontStyles.regular}
    color: ${COLORS.primary1};
    text-decoration: none;
    transition: color 0.2s;

    &:hover {
      color: ${COLORS.primary3};
      text-decoration: none;
    }

    &:active {
      color: ${COLORS.primary3};
      text-decoration: none;
    }
  }

  .k-Form-Checkbox__content {
    flex: 1 0 calc(100% - ${pxToRem(20 + 10)});
    margin: ${pxToRem(5)} 0 0 ${pxToRem(20 + 10)};
  }
`

export const Checkbox = ({
  className,
  id,
  children,
  inputClassName,
  error,
  disabled,
  textProps,
  labelProps,
  onLabelClick,
  indeterminate,
  label,
  ...inputProps
}) => {
  const inputElement = useRef(null)

  useEffect(() => {
    if (inputElement.current != null) {
      inputElement.current.indeterminate = indeterminate
    }
  }, [indeterminate])

  return (
    <StyledCheckbox
      className={classNames(
        'k-Form-Checkbox',
        {
          'k-Form-Checkbox--error': error,
          'k-Form-Checkbox--disabled': disabled,
        },
        className,
      )}
    >
      <input
        ref={inputElement}
        id={id}
        type="checkbox"
        className={classNames('k-Form-Checkbox__input', inputClassName)}
        disabled={disabled}
        {...inputProps}
      />

      {!label && children && (
        <Text
          tag="label"
          htmlFor={id}
          onClick={onLabelClick}
          {...labelProps}
          {...textProps}
          className={classNames(
            'k-Form-Checkbox__label',
            textProps?.className,
            labelProps?.className,
          )}
        >
          {children}
        </Text>
      )}

      {label && !children && (
        <Text
          tag="label"
          htmlFor={id}
          onClick={onLabelClick}
          {...labelProps}
          {...textProps}
          className={classNames(
            'k-Form-Checkbox__label',
            textProps?.className,
            labelProps?.className,
          )}
        >
          {label}
        </Text>
      )}

      {label && children && (
        <>
          <Text
            tag="label"
            htmlFor={id}
            onClick={onLabelClick}
            weight="regular"
            {...labelProps}
            className={classNames(
              'k-Form-Checkbox__label',
              labelProps?.className,
            )}
          >
            {label}
          </Text>
          <Text
            onClick={onLabelClick}
            {...textProps}
            className={classNames(
              'k-Form-Checkbox__content',
              textProps?.className,
            )}
          >
            {children}
          </Text>
        </>
      )}
    </StyledCheckbox>
  )
}

Checkbox.defaultProps = {
  id: 'checkbox',
  textProps: {},
  labelProps: {},
  indeterminate: false,
  error: false,
  disabled: false,
  onLabelClick: () => {},
}

Checkbox.propTypes = {
  textProps: PropTypes.object,
  labelProps: PropTypes.object,
  indeterminate: PropTypes.bool,
  error: PropTypes.bool,
  disabled: PropTypes.bool,
  id: PropTypes.string.isRequired,
  children: PropTypes.node,
  label: PropTypes.string,
  onLabelClick: PropTypes.func,
}
