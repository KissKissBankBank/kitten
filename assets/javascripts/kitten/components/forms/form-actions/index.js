import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { pxToRem } from '../../../helpers/utils/typography'
import { ScreenConfig } from '../../../constants/screen-config'

const StyledFormActions = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;

  .k-FormActions__item {
    margin-left: ${pxToRem(5)};
    margin-right: ${pxToRem(5)};
  }

  &.k-FormActions--spreadOut-s-up {
    @media (min-width: ${pxToRem(ScreenConfig.S.min)}) {
      justify-content: space-between;

      .k-FormActions__item {
        &:first-child {
          margin-left: 0;
        }

        &:last-child {
          margin-right: 0;
        }
      }
    }
  }
`

export const FormActions = ({ className, spreadOutAt, ...others }) => {
  const formActionsClassName = classNames('k-FormActions', className, {
    [`k-FormActions--spreadOut-${spreadOutAt}`]: spreadOutAt,
  })

  return <StyledFormActions className={formActionsClassName} {...others} />
}

FormActions.Item = props => (
  <div
    {...props}
    className={classNames('k-FormActions__item', props.className)}
  />
)

FormActions.propTypes = {
  spreadOutAt: PropTypes.oneOf(['s-up']),
}

FormActions.defaultProps = {
  spreadOutAt: 's-up',
  children: null,
}

FormActions.Item.defaultProps = {
  className: null,
  children: null,
}
