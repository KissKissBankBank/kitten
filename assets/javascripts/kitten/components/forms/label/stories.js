import React from 'react'
import styled from 'styled-components'
import { Label } from './index'
import { DocsPage } from 'storybook/docs-page'
import COLORS from '../../../constants/colors-config'

export default {
  title: 'Forms/Label',
  component: Label,
  parameters: {
    docs: {
      page: () => <DocsPage filepath={__filename} importString="Label" />,
    },
  },
  decorators: [story => <div className="story-Container">{story()}</div>],
  argTypes: {
    focusId: {
      name: 'focusId',
      control: 'text',
    },
    size: {
      name: 'size',
      options: [null, undefined, 'normal', 'tiny', 'micro'],
      control: 'select',
    },
    withoutPointerEvents: {
      name: 'withoutPointerEvents',
      control: 'boolean',
    },
    dot: {
      name: 'dot',
      control: 'object',
    },
  },
}

const args = {
  children: 'Label',
  focusId: null,
  size: 'normal',
  withoutPointerEvents: false,
}

const FocusableComponent = styled.div`
  margin-top: 1.25rem;

  .focused {
    display: none;
  }

  &:focus {
    .focused {
      display: inline;
    }

    .not-focused {
      display: none;
    }
  }
`

export const Default = args => <Label {...args} />

Default.storyName = 'Label'

export const WithDot = args => <Label {...args} />

WithDot.args = {
  ...args,
  dot: {
    width: 7,
    backgroundColor: COLORS.orange,
    title: 'champs obligatoire',
  },
}

export const WithFocusId = args => (
  <>
    <Label {...args} focusId="focusable-component" />
    <FocusableComponent id="focusable-component" tabIndex="0">
      <span className="k-u-color-primary1 k-u-weight-bold k-u-background-color-primary5 focused">
        ðŸ˜¸ Focused
      </span>
      <span className="k-u-color-font2 k-u-weight-regular not-focused">
        ðŸ˜¿ Not focused
      </span>
    </FocusableComponent>
  </>
)

WithFocusId.storyName = 'with focusId prop'
