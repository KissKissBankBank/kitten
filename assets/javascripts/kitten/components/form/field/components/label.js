import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Label } from '../../../form/label'
import { Toggletip } from '../../../information/toggletip'
import { Line } from '../../../structure/line'

export class FieldLabel extends Component {
  render() {
    const {
      children,
      tooltip,
      tooltipId,
      tooltipProps,
      labelProps,
      link,
      tooltipLabel,
      className,
      center,
      ...others
    } = this.props

    return (
      <div
        className={classNames(
          className,
          'k-Field__label',
          'k-u-margin-bottom-single',
        )}
        {...others}
      >
        <Line style={{ lineHeight: 1 }} className={classNames({ 'k-u-flex-justifyContent-center': center })}>
          <Line.Item>
            <Label {...labelProps}>{children}</Label>
          </Line.Item>

          {tooltip && (
            <Line.Item>
              <Toggletip
                id={tooltipId}
                bubbleProps={{ zIndex: 2 }}
                actionLabel="Help"
                {...tooltipProps}
              >
                {tooltip}
              </Toggletip>
            </Line.Item>
          )}

          {link && <Line.Item>{link}</Line.Item>}
        </Line>
      </div>
    )
  }
}

FieldLabel.propTypes = {
  link: PropTypes.node,
  tooltip: PropTypes.string,
  labelProps: PropTypes.object,
  tooltipProps: PropTypes.object,
  center: PropTypes.bool,
}

FieldLabel.defaultProps = {
  tooltip: null,
  labelProps: {},
  center: false,
}
