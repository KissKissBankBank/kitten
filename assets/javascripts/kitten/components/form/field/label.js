import React, { Component } from 'react'
import { Marger } from 'kitten/components/layout/marger'
import { Label } from 'kitten/components/form/label'

export class FieldLabel extends Component {
  render() {
    const { children, ...others } = this.props

    return (
      <Marger bottom="1.5">
        <Label {...others} margin={false} size="tiny">
          {children}
        </Label>
      </Marger>
    )
  }
}
