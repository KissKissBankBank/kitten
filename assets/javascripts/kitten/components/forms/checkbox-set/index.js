import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import classNames from 'classnames'
import { Checkbox } from '../../../components/forms/checkbox'
import { pxToRem } from '../../../helpers/utils/typography'
import { Label } from '../../../components/forms/label'

const StyledCheckboxSet = styled.fieldset`
  margin: 0;
  padding: 0;
  border: 0;
  display: flex;
  flex-direction: column;
  line-height: 1.3rem;

  & > legend {
    padding: 0;
  }

  .k-Form-CheckboxSet__label {
    margin-bottom: ${pxToRem(10)};
  }

  .k-Form-CheckboxSet__checkbox {
    margin: ${pxToRem(5)} 0;

    &:first-of-type {
      margin-top: 0;
    }
    &:last-of-type {
      margin-bottom: 0;
    }
  }
`

export const CheckboxSet = ({
  items,
  disabled,
  className,
  name,
  error,
  label,
  children,
  ...props
}) => (
  <StyledCheckboxSet
    className={classNames('k-Form-CheckboxSet', className)}
    disabled={disabled}
    {...props}
  >
    {label && (
      <Label tag="legend" className="k-Form-CheckboxSet__label">
        {label}
      </Label>
    )}
    {children && !label && <legend>{children}</legend>}
    {items.map(({ id, className, ...itemProps }) => (
      <Checkbox
        id={id}
        error={error}
        name={name}
        key={id}
        {...itemProps}
        className={classNames('k-Form-CheckboxSet__checkbox', className)}
      />
    ))}
  </StyledCheckboxSet>
)

CheckboxSet.propTypes = {
  name: PropTypes.string.isRequired,
  error: PropTypes.bool,
  label: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      id: PropTypes.string.isRequired,
      defaultChecked: PropTypes.bool,
      children: PropTypes.node,
    }),
  ),
  disabled: PropTypes.bool,
}

CheckboxSet.defaultProps = {
  name: 'checkboxSet',
  error: false,
  label: null,
  items: [
    {
      label: null,
      children: null,
      defaultChecked: true,
      id: 'myCheckbox', // Replace by a real value
    },
  ],
  disabled: false,
}
