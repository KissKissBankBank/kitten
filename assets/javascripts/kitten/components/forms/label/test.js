import React from 'react'
import renderer from 'react-test-renderer'
import { Label } from '../../forms/label'

describe('<Label />', () => {
  it('should match its empty snapshot', () => {
    const tree = renderer.create(<Label />).toJSON()

    expect(tree).toMatchSnapshot()
  })
  it('should match its propped snapshot', () => {
    const tree = renderer
      .create(
        <Label
          size="tiny"
          children="Custom form label"
          focusId="customId"
          withoutPointerEvents
          tag="span"
          dot={{
            width: 7,
            height: 7,
            backgroundColor: '#ff7800',
            title: 'champs obligatoire',
          }}
        />,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
