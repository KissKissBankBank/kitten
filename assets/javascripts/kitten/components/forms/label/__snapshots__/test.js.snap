// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label /> should match its empty snapshot 1`] = `
<label
  className="sc-bdVaJa hEefBW k-Label k-Label--normal"
  onClick={[Function]}
>
  Label
</label>
`;

exports[`<Label /> should match its propped snapshot 1`] = `
<span
  className="sc-bdVaJa hEefBW k-Label k-Label--tiny k-Label--withoutPointerEvents"
  htmlFor="customId"
  onClick={[Function]}
>
  Custom form label
  <span
    className="k-Label--dot"
    style={
      Object {
        "--dot-background-color": "#ff7800",
        "--dot-width": "0.4375rem",
      }
    }
    tabIndex="-1"
    title="champs obligatoire"
  />
</span>
`;
